tasks/update-mlnx-ofed.yml

```
@@ -0,0 +1,19 @@
+---
+ - name: Add script content
+   ansible.builtin.copy:
+     dest: /etc/kernel/header_postinst.d/01-dkms-mlnx-ofed-kernel
+     mode: 750
+     content: |
+       #!/bin/bash
+
+       # We're passed the version of the kernel being installed
+       inst_kern=$1
+
+       uname_m=$(uname -m)
+
+       # Install mlnx-ofed-kernel first
+       for mod in $(dkms status mlnx-ofed-kernel | cut -d, -f1 | sort -u); do
+           /usr/sbin/dkms install ${mod} -k ${inst_kern}
+       done
```

Why `uname -m` and `dkms status` but `/usr/sbin/dkms` and `/usr/bin/update-alternatives`?

And is the idea here to reinstall all existing versions of `mlnx-ofed-kernel` for a new kernel? Is there a chance the rebuild of such out-of-tree modules could break after a kernel update, and if so, what happens? Is this script supposed to continue, or exit with an error code, or what?

---

Indeed absolute path are not mandatory. This is inspired by the script `/etc/kernel/header_postinst.d/dkms` shipped with the dkms package.

> is the idea here to reinstall all existing versions of mlnx-ofed-kernel for a new kernel?

Yes, like the script above does. This new script should execute **before** `/etc/kernel/header_postinst.d/dkms`, so the mlnx ofed modules will be available for the new kernel when dkms will try to build other modules that depends on it (e.g. Lustre). With the default script, `lustre-client` modules are built before `mlnx-ofed` modules, using the wrong version of `mlnx-ofed` and the wrong symbols.

> Is there a chance the rebuild of such out-of-tree modules could break after a kernel update, and if so, what happens?

That could break indeed. I guess the script will fail and `apt upgrade` will report a failure. Do you believe we should catch the rc of `dkms install`?

---

I wasn't even aware of `/etc/kernel/header_postinst.d/` until today. So I don't exactly know the protocol for what exit codes signal and to whom.

But if this script ignores errors then I fear that you may silently end up in exactly the scenario you describe:

> This new script should execute before /etc/kernel/header_postinst.d/dkms, so the mlnx ofed modules will be available for the new kernel when dkms will try to build other modules that depends on it (e.g. Lustre). With the default script, lustre-client modules are built before mlnx-ofed modules, using the wrong version of mlnx-ofed and the wrong symbols.

so it may be better to be noisy and die by using `set -e`. Provided, that is, that whoever runs this postinstall script _also_ aborts upon failure.

---

We could add more verbosity to the script in case of errors. However, this script will be used when `infiniteband_dkms_postinst` is true, so during a kernel update. The monitoring system will be provided with information if these modules did not build correctly. Some file systems will fail to mount.

---

I ran some tests, if the script should fail, the apt upgrade would fail as well. So we will know.
I guess this is fine as it is, as close a possible to similar scripts shipped with the distro.